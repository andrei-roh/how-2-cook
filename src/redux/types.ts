export const SET_USER = 'SET_USER';
export const SET_LOADING = 'SET_LOADING';
export const SET_ERROR = 'SET_ERROR';
export const ADD_RECIPES_TO_LIST = 'ADD_RECIPES_TO_LIST';
export const UPDATE_RECIPES_LIST = 'UPDATE_RECIPES_LIST';
export const ADD_IMAGES_TO_LIST = 'ADD_IMAGES_TO_LIST';
export const SET_RECIPE_SEARCH_INPUT = 'SET_RECIPE_SEARCH_INPUT';
export const SET_MANAGE_RECIPES_SCROLL_SIZE = 'SET_MANAGE_RECIPES_SCROLL_SIZE';
export const SET_MANAGE_RECIPES_SCROLL_DIRECTION =
  'SET_MANAGE_RECIPES_SCROLL_DIRECTION';
export const SET_SHOW_MENU = 'SET_SHOW_MENU';
export const ADD_INGREDIENTS_TO_LIST = 'ADD_INGREDIENTS_TO_LIST';
export const SET_INGREDIENT_SEARCH_INPUT = 'SET_INGREDIENT_SEARCH_INPUT';
export const UPDATE_INGREDIENTS_LIST = 'UPDATE_INGREDIENTS_LIST';


