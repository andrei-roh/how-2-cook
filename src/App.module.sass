@use 'src/style/colors'

.app 
  padding-top: 46px
  background-color: colors.$white
  height: 100vh
  box-sizing: border-box
  display: flex
  flex-direction: column
  align-items: center

