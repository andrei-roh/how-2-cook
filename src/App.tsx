import { Outlet } from 'react-router-dom';
import css from './App.module.sass';
import { Header, Notification } from './components';

const App = () => {
  return (
    <>
      <Header />
      <div className={css.app}>
        <Outlet />
      </div>
      <Notification />
    </>
  );
};

export default App;
