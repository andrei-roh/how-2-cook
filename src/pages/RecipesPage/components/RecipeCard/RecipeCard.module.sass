@use 'src/style/colors'
@import 'src/style/typography'

.recipeCardWrapper
    display: flex
    flex-direction: row
    justify-content: space-between
    width: 100%
    box-sizing: border-box
    border: 2px solid colors.$gray
    border-radius: 12px

    .recipeContainer
        width: 100%
        display: grid
        grid-template-columns: 50% 50%
        align-items: center
        justify-content: space-between

        .recipeCardLoader
            width: 100%
            height: 179px
            display: flex
            justify-content: center
            align-items: center

        .recipeCardImage
            width: 100%
            height: calc(100vw / 2 - 18px)
            border-top-left-radius: 10px
            border-bottom-left-radius: 10px
            object-fit: cover

        .recipeData
            display: flex
            flex-direction: column
            justify-content: space-around
            width: 100%
            height: 100%
            text-align: center

            .recipeName
                @extend .H6
                align-self: center

            .recipeType
                @extend .T3
                align-self: center

            .buttonsPanel
                display: flex
                flex-direction: row
                justify-content: space-around
    
    .recipeCardLogo
        width: 25px

.recipeEditButton
    align-self: flex-start
