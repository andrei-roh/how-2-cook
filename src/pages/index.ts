export { TitlePage } from './TitlePage/TitlePage';
export { RecipesPage } from './RecipesPage/RecipesPage';
export { CreateRecipePage } from './CreateRecipePage/CreateRecipePage';
export { ShowRecipePage } from './ShowRecipePage/ShowRecipePage';
export { EditRecipePage } from './EditRecipePage/EditRecipePage';
export { IngredientsPage } from './IngredientsPage/IngredientsPage';
export { WeekPlanningPage } from './WeekPlanningPage/WeekPlanningPage';
export { MyPlansPage } from './MyPlansPage/MyPlansPage';