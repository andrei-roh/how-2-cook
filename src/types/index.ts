import { Action } from '@reduxjs/toolkit';

export interface IUser {
  email: string;
}

export enum DishType {
  Main = 'main',
  Soup = 'soup',
  Dessert = 'dessert',
  Salat = 'salat',
  Snack = 'snack',
}

export type RecipeImage = {
  [key: string]: string;
};

export interface IIngredient {
  id: string;
  name: string;
}

export interface IRecipe {
  id: string;
  imageUrl: string;
  name: string;
  type: DishType;
  isVegan: boolean;
  ingredients: string[];
  description: string;
  composition?: string;
  createdAt?: string;
  createdBy?: string;
  updatedAt?: string;
  updatedBy?: string;
}

export interface IPlanningList {
  id: string;
  name: string;
  ingredientsList: string[];
  recipesList: string[];
}

export interface IState {
  user: IUser;
  loading: boolean;
  error: null | string;
  recipesList: IRecipe[];
  imagesList: RecipeImage[];
  recipesPage: {
    recipeSearchInput: string;
    scrollSize: number;
    scrollDirection: ScrollDirection;
  };
  isMenu: boolean;
  ingredientsList: IIngredient[];
  ingredientsPage: {
    ingredientSearchInput: string;
  }
  previousRoute: string;
  planningPage: {
    ingredientsList: string[];
    recipesList: IRecipe[];
  }
}

export interface IAction extends Action {
  type: string;
  payload: never;
}

export interface IValidationError {
  [key: string]: string;
}

export enum FirebaseErrorType {
  InvalidEmail = 'auth/invalid-email',
  InvalidPassword = 'auth/invalid-credential',
  TooManyRequsts = 'auth/too-many-requests',
}

export enum ImageType {
  Webp = 'image/webp',
  Jpeg = 'image/jpeg',
  Jpg = 'image/jpg',
  Png = 'image/png',
}

export interface IOption {
  name: string;
  value: string;
}

export enum ScrollDirection {
  Up = 'up',
  Down = 'down',
}

export enum Severity {
  Success = 'success',
  Info = 'info',
  Warning = 'warning',
  Error = 'error',
}

export interface ICurrentRecipe extends Partial<IRecipe> {}
