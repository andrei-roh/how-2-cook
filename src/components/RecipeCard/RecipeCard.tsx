import css from './RecipeCard.module.sass';
import { IRecipe, IState } from 'src/types';
import Edit from 'src/assets/pencil.svg';
import Show from 'src/assets/eye.svg';
import Vegetarian from 'src/assets/vegetarian.svg';
import Bin from 'src/assets/paper-bin.svg';
import { useNavigate } from 'react-router-dom';
import { getRecipeImage } from 'src/utils/getRecipeImage';
import { Dispatch } from '@reduxjs/toolkit';
import { useDispatch, useSelector } from 'react-redux';
import { addImagesToList, updateRecipesList } from 'src/redux/actions';
import { DISH_TYPE } from 'src/constants';
import { useEffect, useState } from 'react';
import CircularProgress from '@mui/material/CircularProgress';
import Stack from '@mui/material/Stack';
import Typography from '@mui/material/Typography';
import Card from '@mui/material/Card';
import { Modal } from 'src/components';
import { deleteRecipe } from 'src/utils/deleteRecipe';

interface RecipeCardProps extends IRecipe {
  isControlled?: boolean;
}

export const RecipeCard = ({
  id,
  imageUrl,
  name,
  type,
  isVegan,
  isControlled = false,
}: RecipeCardProps) => {
  const navigate = useNavigate();
  const dispatch: Dispatch = useDispatch();
  const imagesList = useSelector((state: IState) => state.imagesList);
  const recipesList = useSelector((state: IState) => state.recipesList);
  const [isDelete, setIsDelete] = useState(false);
  const currentImageUrl = imagesList.find(
    (image) => image.id === imageUrl
  )?.value;

  const handleEditRecipe = () => {
    navigate(`/recipe/edit/${id}`);
  };

  const handleShowRecipe = () => {
    navigate(`/recipe/show/${id}`);
  };

  const handleShowModal = () => setIsDelete(true);
  const handleCloseModal = () => setIsDelete(false);

  const handleDeleteRecipe = () => {
    deleteRecipe(id, imageUrl, name).then((result) => {
      if (result) {
        const deletedRecipeIndex = recipesList.findIndex(
          (recipe) => recipe.id === id
        );
        const updatedRecipesList: IRecipe[] = [...recipesList];
        updatedRecipesList.splice(deletedRecipeIndex, 1);

        dispatch(updateRecipesList([...updatedRecipesList]));
        handleCloseModal();
      }
    });
  };

  useEffect(() => {
    if (!currentImageUrl) {
      getRecipeImage(imageUrl).then((url) => {
        if (url) {
          dispatch(addImagesToList([{ id: imageUrl, value: url }]));
        }
      });
    }
  }, [currentImageUrl, dispatch, id, imageUrl]);

  return (
    <Card variant='elevation' elevation={3} className={css.recipeCardWrapper}>
      <Stack className={css.recipeContainer}>
        {currentImageUrl ? (
          <img className={css.recipeCardImage} id={id} src={currentImageUrl} />
        ) : (
          <div className={css.recipeCardLoader}>
            <CircularProgress />
          </div>
        )}
        <Stack className={css.recipeData}>
          <Typography className={css.recipeName} noWrap>
            {name}
          </Typography>
          <Typography className={css.recipeType}>
            {DISH_TYPE.find(({ value }) => value === type)?.name}
            {isVegan && <img width={12} src={Vegetarian} />}
          </Typography>
          <Stack className={css.buttonsPanel}>
            {isControlled && (
              <img
                onClick={handleEditRecipe}
                className={css.recipeCardLogo}
                src={Edit}
                alt='Edit Recipe Button'
              />
            )}
            <img
              onClick={handleShowRecipe}
              className={css.recipeCardLogo}
              src={Show}
              alt='Show Recipe Button'
            />
            {isControlled && (
              <img
                onClick={handleShowModal}
                className={css.recipeCardLogo}
                src={Bin}
                alt='Delete Recipe Button'
              />
            )}
          </Stack>
        </Stack>
        {isDelete && (
          <Modal
            cancelButtonMessage='Отмена'
            submitButtonMessage='Удалить'
            isLoading={false}
            handleClose={handleCloseModal}
            handleSubmit={handleDeleteRecipe}
            message={`Вы уверены, что хотите удалить рецепт ${name}?`}
          />
        )}
      </Stack>
    </Card>
  );
};
