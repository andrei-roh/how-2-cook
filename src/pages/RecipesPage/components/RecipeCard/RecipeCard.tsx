import css from './RecipeCard.module.sass';
import { IRecipe, IState } from 'src/types';
import Edit from 'src/assets/pencil.svg';
import Show from 'src/assets/eye.svg';
import Vegetarian from 'src/assets/vegetarian.svg';
import { useNavigate } from 'react-router-dom';
import { getRecipeImage } from 'src/utils/getRecipeImage';
import { Dispatch } from '@reduxjs/toolkit';
import { useDispatch, useSelector } from 'react-redux';
import { addImagesToList } from 'src/redux/actions';
import { DISH_TYPE } from 'src/constants';
import { useEffect } from 'react';
import CircularProgress from '@mui/material/CircularProgress';
import Stack from '@mui/material/Stack';
import Typography from '@mui/material/Typography';
import Card from '@mui/material/Card';

interface RecipeCardProps extends IRecipe {}

export const RecipeCard = ({
  id,
  imageUrl,
  name,
  type,
  isVegan,
}: RecipeCardProps) => {
  const navigate = useNavigate();
  const dispatch: Dispatch = useDispatch();
  const imagesList = useSelector((state: IState) => state.imagesList);
  const currentImageUrl = imagesList.find(
    (image) => image.id === imageUrl
  )?.value;

  const handleEditRecipe = () => {
    navigate(`/recipe/edit/${id}`);
  };

  const handleShowRecipe = () => {
    navigate(`/recipe/show/${id}`);
  };

  useEffect(() => {
    if (!currentImageUrl) {
      getRecipeImage(imageUrl).then((url) => {
        if (url) {
          dispatch(addImagesToList([{ id: imageUrl, value: url }]));
        }
      });
    }
  }, [currentImageUrl, dispatch, id, imageUrl]);

  return (
    <Card className={css.recipeCardWrapper}>
      <Stack className={css.recipeContainer}>
        {currentImageUrl ? (
          <img className={css.recipeCardImage} id={id} src={currentImageUrl} />
        ) : (
          <div className={css.recipeCardLoader}>
            <CircularProgress />
          </div>
        )}
        <Stack className={css.recipeData}>
          <Typography className={css.recipeName} noWrap>
            {name}
          </Typography>
          <Typography className={css.recipeType}>
            {DISH_TYPE.find(({ value }) => value === type)?.name}
            {isVegan && <img width={12} src={Vegetarian} />}
          </Typography>
          <Stack className={css.buttonsPanel}>
            <img
              onClick={handleEditRecipe}
              className={css.recipeCardLogo}
              src={Edit}
              alt='Edit Recipe Button'
            />
            <img
              onClick={handleShowRecipe}
              className={css.recipeCardLogo}
              src={Show}
              alt='Show Recipe Button'
            />
          </Stack>
        </Stack>
      </Stack>
    </Card>
  );
};
