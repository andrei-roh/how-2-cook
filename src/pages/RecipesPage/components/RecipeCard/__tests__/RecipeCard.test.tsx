import { describe, expect, it, vi } from 'vitest';
import { render } from '@testing-library/react';
import { DishType, IRecipe, IUser } from 'src/types';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import { RecipeCard } from '../RecipeCard';

const navigateMock = vi.fn();
const dispatchMock = vi.fn();
const userMock = { email: 'Test@email.com' };
const imagesListMock = [
  {
    id: 'test_image_src_1',
    value: 'test/image/src/1',
  },
];
const recipeMock: IRecipe = {
  id: '00000000-0000-0000-0000-000000000001',
  imageUrl: 'test_image_src_1',
  name: 'Test Recipe 1',
  type: DishType.Fish,
  ingredients: 'Test Ingredients 1',
  description: 'Test Description 1',
};

const stateMock = {
  user: userMock as IUser,
  loading: false,
  error: null,
  imagesList: imagesListMock,
};

const reducerMock = (state = stateMock) => {
  return state;
};

const storeMock = configureStore({ reducer: reducerMock });

vi.mock('react-router-dom', async () => {
  const actual = await vi.importActual('react-router-dom');

  return {
    ...actual,
    useNavigate: () => ({ navigate: navigateMock }),
  };
});

vi.mock('react-redux', async () => {
  const actual = await vi.importActual('react-redux');

  return {
    ...actual,
    useDispatch: () => ({ dispatch: dispatchMock }),
  };
});

vi.mock('src/utils/getRecipeImage');

describe('Testing: RecipeCard', () => {
  it('should render', () => {
    const { container } = render(
      <Provider store={storeMock}>
        <RecipeCard {...recipeMock} />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
