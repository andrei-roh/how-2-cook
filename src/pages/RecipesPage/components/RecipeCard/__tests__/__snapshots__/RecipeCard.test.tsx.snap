// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing: RecipeCard > should render 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root _recipeCardWrapper_a06f98 css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiStack-root _recipeContainer_a06f98 css-nen11g-MuiStack-root"
    >
      <img
        class="_recipeCardImage_a06f98"
        id="00000000-0000-0000-0000-000000000001"
        src="test/image/src/1"
      />
      <div
        class="MuiStack-root _recipeData_a06f98 css-nen11g-MuiStack-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap _recipeName_a06f98 css-fvtpii-MuiTypography-root"
        >
          Test Recipe 1
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1 _recipeType_a06f98 css-ahj2mt-MuiTypography-root"
        >
          Основное
        </p>
        <div
          class="MuiStack-root _buttonsPanel_a06f98 css-nen11g-MuiStack-root"
        >
          <img
            alt="Edit Recipe Button"
            class="_recipeCardLogo_a06f98"
            src="/src/assets/pencil.svg"
          />
          <img
            alt="Show Recipe Button"
            class="_recipeCardLogo_a06f98"
            src="/src/assets/eye.svg"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
